### PR TITLE
4228 keycloak integration login register password reset ux

### PR DESCRIPTION
**Pull request by collinpreston**
#### Pre-Flight checklist
- [x] Testing
  - [x] Changes have been manually tested
- [x] Settings
  - [x] New settings are documented and present in `app.json`
  - [x] New settings have reasonable development defaults, if applicable

#### What are the relevant tickets?
https://github.com/mitodl/open-discussions/issues/4228

#### What's this PR do?
When the environment variable `FEATURE_KEYCLOAK_ENABLED` is set to True, all password, registration, and login functionality should be handled by Keycloak.  The default value for `FEATURE_KEYCLOAK_ENABLED` is False.

When `FEATURE_KEYCLOAK_ENABLED` is `True`:
- URLs no longer needed should be disabled.
- Login and Sign up UI and URLs should initiate the Keycloak login workflow.
- Password reset URL's and UI should initiate the Keycloak password change workflow.
- Logging out of Open Discussions should terminate the user's Keycloak session.

When `FEATURE_KEYCLOAK_ENABLED` is `False`: 
- All functionality added by this PR should be disabled.

3 environment variables have been added and are required:

- `FEATURE_KEYCLOAK_ENABLED`
- `KEYCLOAK_BASE_URL`
- `KEYCLOAK_REALM_NAME`

## How should this be manually tested?
These instructions assume that you have Keycloak in the QA environment configured with your local instance of Open Discussions.

- Add `KEYCLOAK_BASE_URL=https://sso-qa.odl.mit.edu` to your .env file.
- Add `KEYCLOAK_REALM_NAME=olapps` to your .env file.
### Testing URLs
1. In the `.env` file, add `FEATURE_KEYCLOAK_ENABLED=True`.
2. The following URL endpoints should redirect to Keycloak's login screen: http://od.odl.local:8063/login/, http://od.odl.local:8063/signup/.
3. The following URL endpoints should not respond: http://od.odl.local:8063/password_reset/, http://od.odl.local:8063/signup/confirm/, http://od.odl.local:8063/password_reset/confirm/, http://od.odl.local:8063/api/v0/login/email/, http://od.odl.local:8063/api/v0/login/password/, http://od.odl.local:8063/api/v0/register/email/, http://od.odl.local:8063/api/v0/register/confirm/, http://od.odl.local:8063/api/v0/register/details/, http://od.odl.local:8063/api/v0/password_reset/, http://od.odl.local:8063/api/v0/password_reset/confirm/, http://od.odl.local:8063/api/v0/set_password/
4. The following API endpoints should be accessible using http://od.odl.local:8063/api/v0/auth/<EMAIL> and http://od.odl.local:8063/api/v0/update_password_request/.

### Testing Login UI
1. In the `.env` file, add `FEATURE_KEYCLOAK_ENABLED=True`.
2. The following UI elements should redirect to Keycloak's login screen: 
- Login option from top right menu.  
- Login option shown when clicking "Create a post" from the home screen.
3. Login using one of the options from step 2. 

### Testing Password Change
1. In the `.env` file, add `FEATURE_KEYCLOAK_ENABLED=True`.
2. Follow the steps in "Testing Login UI".
3. Navigate to http://od.odl.local:8063/settings/password which should redirect you to the Keycloak password change screen.  There should also be a link at the top-right corner of the page which allows you to navigate back to Open Discussions.

### Testing Logout
1. In the `.env` file, add `FEATURE_KEYCLOAK_ENABLED=True`
2. In a browser tab, follow the steps in "Testing Login UI".
4. In the Open discussions browser tab, click "Sign Out" from the top-right menu of Open Discussions or visit http://od.odl.local:8063/logout.
5. You should be redirected to a Keycloak page asking you to confirm your logout.  Confirm the logout.
6. Visit http://od.odl.local:8063/ and confirm you are logged out of Open Discussions.
7. Visit https://sso-qa.odl.mit.edu/realms/olapps/account/ and confirm you are logged out of Keycloak.

### Testing the environment variable as `False`.
1. In the `.env` file, add `FEATURE_KEYCLOAK_ENABLED=False`.
9. Restart the application.
10. All URL endpoints and UI related to passwords, registration, and login functionality should perform as they were before this PR.
11. The following API endpoints should NOT be accessible using http://od.odl.local:8063/api/v0/auth/<EMAIL> and http://od.odl.local:8063/api/v0/update_password_request/.

### Testing default environment variable value
1. In the `.env` file, remove `FEATURE_KEYCLOAK_ENABLED=False`.
6. Restart the application.
8. All URL endpoints and UI related to passwords, registration, and login functionality should perform as they were before this PR.
9. The following API endpoints should NOT be accessible using http://od.odl.local:8063/api/v0/auth/<EMAIL> and http://od.odl.local:8063/api/v0/update_password_request/.

### CODE REVIEW

**Review by mbertrand — COMMENTED** [review]
A couple functionality-related things I noticed, this was a few commits ago so things may have changed:

- If I log out, then press the login link, I automatically get logged back in as the same user.  Some people might have different accounts (like I do for testing) - seems like I'd have to clear cookies to log in under another account?
- The following url also logs me me back in after logging out, but it is in the list that is not supposed to work anymore: http://od.odl.local:8063/signup/confirm/
- Not sure if related to this PR, but the account settings page looks like this:
Screenshot 2023-10-11 at 11 19 25 AM
- I never got a password reset email after going to http://od.odl.local:8063/settings/password

**Comment by collinpreston** [comment]
> A couple functionality-related things I noticed, this was a few commits ago so things may have changed:
> 
> * If I log out, then press the login link, I automatically get logged back in as the same user.  Some people might have different accounts (like I do for testing) - seems like I'd have to clear cookies to log in under another account?
> * The following url also logs me me back in after logging out, but it is in the list that is not supposed to work anymore: http://od.odl.local:8063/signup/confirm/
> * Not sure if related to this PR, but the account settings page looks like this:
>   Screenshot 2023-10-11 at 11 19 25 AM
> * I never got a password reset email after going to http://od.odl.local:8063/settings/password

Items 1, 2, and 4 should be fixed with the latest commits.  Item 3 will be done in another issue: https://github.com/mitodl/open-discussions/issues/4229

File: open_discussions/settings.py [hunk 1]
```diff
@@ -390,7 +400,7 @@
 )
 SOCIAL_AUTH_OL_OIDC_OIDC_ENDPOINT = get_string(
     name="SOCIAL_AUTH_OL_OIDC_OIDC_ENDPOINT",
-    default=None,
+    default="https://www.keycloak_base_url.edu/realms/realm_name",
```
Comment (mbertrand) [inline open_discussions/settings.py:403]: Should this make use of `KEYCLOAK_BASE_URL` above?

```python
 default=f"{KEYCLOAK_BASE_URL}/realms/realm_name",
```

File: .env.example [hunk 2]
```diff
@@ -24,4 +24,15 @@ SOCIAL_AUTH_MICROMASTERS_LOGIN_URL=
...
+SOCIAL_AUTH_OL_OIDC_OIDC_ENDPOINT=https://sso-qa.odl.mit.edu/realms/olapps
+OIDC_ENDPOINT=https://sso-qa.odl.mit.edu/realms/olapps
+SOCIAL_AUTH_OL_OIDC_KEY=ol-open-discussions-local
+SOCIAL_AUTH_OL_OIDC_SECRET=tpaVPFg4OAz0AVuQ61CGRVLqOGPP2v4L
+AUTHORIZATION_URL=https://sso-qa.odl.mit.edu/realms/olapps/protocol/openid-connect/auth
+ACCESS_TOKEN_URL=https://sso-qa.odl.mit.edu/realms/olapps/protocol/openid-connect/token
+USERINFO_URL=https://sso-qa.odl.mit.edu/realms/olapps/protocol/openid-connect/userinfo
+
+KEYCLOAK_BASE_URL=https://sso-qa.odl.mit.edu
+KEYCLOAK_REALM_NAME=olapps
```
Comment (mbertrand) [inline .env.example:38]: Can you add a README section on how to enable keycloak and set these values?  Also, a working `SOCIAL_AUTH_OL_OIDC_SECRET` shouldn't be included here, just noticed I got a github alert email about it.   Should probably change it to a new value now.

File: authentication/views.py [hunk 3]
```diff
@@ -288,3 +291,75 @@ def get_user_details_for_keycloak(request, email):
...
+    """
+
+    def post(self, request, *args, **kwargs):
+        """
+        POST endpoint for logging a user out.
+        Raises 404 if the user is not included in the request.
+        """
+        user = getattr(request, "user", None)
+        if user:
+            logout_of_keycloak(user)
```
Comment (mbertrand) [inline authentication/views.py:350]: Logout seems to work, but I still get logged in automatically as the same user again when I click the sign-in link, no prompt for username/password

File: authentication/urls.py [hunk 4]
```diff
@@ -1,63 +1,80 @@
...
-    CustomDjoserAPIView,
     get_user_details_for_keycloak,
+    login_complete,
+    post_request_password_update,
 )
-
+from open_discussions.features import KEYCLOAK_ENABLED
 
 urlpatterns = [
-    re_path(r"^api/v0/login/email/$", LoginEmailView.as_view(), name="psa-login-email"),
```
Comment (mbertrand) [inline authentication/urls.py:19]: Should we keep some or all of these around if `FEATURE_KEYCLOAK_ENABLED=False`?

```python
if not settings.FEATURE_KEYCLOAK_ENABLED:
    urlpatterns.extend([
        re_path(r"^api/v0/login/email/$", LoginEmailView.as_view(), name="psa-login-email"),
        etc
    ])
```

Similar to https://github.com/mitodl/open-discussions/pull/4234/files#diff-da53844e0b6a541573b96bec40d2b104890c238dc460ae98c829944bde64b51aR103-R120

File: frontends/open-discussions/src/lib/url.js [hunk 5]
```diff
@@ -97,21 +98,37 @@ export const CONTENT_POLICY_URL = "/content_policy/"
...
 export const PASSWORD_CHANGE_URL = "/settings/password"
 
 // auth urls
-export const LOGIN_URL = "/login/"
-export const LOGIN_PASSWORD_URL = "/login/password/"
-export const LOGIN_PROVIDER_URL = "/login/external/"
-
-export const REGISTER_URL = "/signup/"
-export const REGISTER_CONFIRM_URL = "/signup/confirm/"
-export const REGISTER_DETAILS_URL = "/signup/details/"
```
Comment (mbertrand) [inline frontends/open-discussions/src/lib/url.js:110]: Keep around and use instead of the ones below if `FEATURE_KEYCLOAK_ENABLED=False`?

**Review by mbertrand — CHANGES_REQUESTED** [review]

File: authentication/urls.py (hunk 4 above)
Comment (collinpreston) [reply authentication/urls.py:19]: I have a bunch down under the `else` condition.  Do you see if I'm missing one?

**Review by mbertrand — APPROVED** [review]
👍

**Pull request merged**